Add tech preview documentation for deploying models in OCI containers (#517)
opendatahub-io/opendatahub-documentation: modules/using-oci-containers-for-model-storage.adoc

COPY --chown=0:0 models /models
RUN chmod -R a=rX /models

# nobody user
USER 65534 
----
+
[NOTE]
====
* In this example, `ubi8-micro` is used as a base container image. You cannot use empty images that do not provide a shell, such as `scratch`, because KServe uses the shell to ensure the model files are accessible to the model server. 
* Ownership of the copied model files and read permissions are granted to the `root` group. OpenShift runs containers with a random user ID and the `root` group ID. Changing ownership of the group ensures that the model server can access them.
====
+
. Confirm that the models follow the directory structure shown using the `tree` command:
+
[source]
----
tree

.
├── Containerfile
└── models
    └── 1
        └── mobilenetv2-7.onnx
----
+
. Create the OCI container image with Podman, and upload it to a registry. For
example, using Quay as the registry:
+
[source]
----
podman build --format=oci -t quay.io/<user_name>/<repository_name>:<tag_name> .
podman push quay.io/<user_name>/<repository_name>:<tag_name>
----
+
[NOTE]
====
If your repository is private, ensure you are authenticated to the registry before uploading your container image.
====

> == Deploying a model stored in an OCI image in a public repository

Reviewer: This header ends with `in a public repository`, while the next one ends with `from a private repository`. 

It is `in` vs `from`. I don't know which one is grammatically correct :-) If both are right, should we change the writing of one to be consistent?

[NOTE]
====
By default in KServe, models are exposed outside the cluster and not protected with authorization. 
====

. Create a namespace to deploy the model:
+
[source]
----
oc new-project oci-model-example
----
+
. Use the OpenData Hub project `kserve-ovms` template to create a `ServingRuntime` resource and configure the OpenVINO model server in the new namespace:
+
[source]
----
oc process -n opendatahub -o yaml kserve-ovms | oc apply -f -
----
+
--
.. Verify that the `ServingRuntime` has been created with the `kserve-ovms` name:
+
[source]
----
oc get servingruntimes

NAME          DISABLED   MODELTYPE     CONTAINERS         AGE
kserve-ovms              openvino_ir   kserve-container   1m
----
--
+
. Create an `InferenceService` YAML resource with the following values:
+
[source]
----
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  name: sample-isvc-using-oci
spec:
  predictor:
    model:
      runtime: kserve-ovms # Ensure this matches the name of the ServingRuntime resource
      modelFormat:
        name: onnx
      storageUri: oci://quay.io/<user_name>/<repository_name>:<tag_name>
----
+

> [NOTE]

Reviewer: FWIW, this is more on the warning side, than a simple note.

Author: I bumped it up to an "Important" tag :)

* Ownership of the copied model files and read permissions are granted to the `root` group. OpenShift runs containers with a random user ID and the `root` group ID. Changing ownership of the group ensures that the model server can access them.
====
+
. Confirm that the models follow the directory structure shown using the `tree` command:
+
[source]
----
tree

.
├── Containerfile
└── models
    └── 1
        └── mobilenetv2-7.onnx
----
+
. Create the OCI container image with Podman, and upload it to a registry. For
example, using Quay as the registry:
+
[source]
----
podman build --format=oci -t quay.io/<user_name>/<repository_name>:<tag_name> .
podman push quay.io/<user_name>/<repository_name>:<tag_name>
----
+
[NOTE]
====
If your repository is private, ensure you are authenticated to the registry before uploading your container image.
====

== Deploying a model stored in an OCI image from a public repository

[NOTE]
====
By default in KServe, models are exposed outside the cluster and not protected with authorization. 
====

. Create a namespace to deploy the model:
+
[source]
----
oc new-project oci-model-example
----
+

> . Use the OpenData Hub project `kserve-ovms` template to create a `ServingRuntime` resource and configure the OpenVINO model server in the new namespace:

Author: Do we need to have "OpenData Hub" conditionalized to show RHOAI if on docs.redhat.com? @israel-hdez

Reviewer 2: change to `OpenData Hub`  to  `redhat-ods-applications`

Reviewer: So, on docs.redhat.com it should show as "RHOAI Applications project". If these docs also go to https://opendatahub.io/docs, I think we need the conditional.

Author: Thanks! I added the conditional back in with your suggestions.

sample-isvc-using-oci   https://sample-isvc-using-oci-oci-model-example.example   True           100                              sample-isvc-using-oci-predictor-00001   1m
----

== Deploying a model stored in an OCI image from a private repository

To deploy a model stored from a private OCI repository, you must configure an image pull secret. For more information about creating an image pull secret, see link:https://docs.openshift.com/container-platform/latest/openshift_images/managing_images/using-image-pull-secrets.html[Using image pull secrets^].

. Follow the steps in the previous section for deploying a model. However, when creating the `InferenceService` in step 3, specify your pull secret in the
`spec.predictor.imagePullSecrets` field:
+
[source]
----
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  name: sample-isvc-using-private-oci
spec:
  predictor:
    model:
      runtime: kserve-ovms
      modelFormat:
        name: onnx
      storageUri: oci://quay.io/<user_name>/<repository_name>:<tag_name>

>     imagePullSecrets: # Specify image pull secrets to use for fetching container images (including OCI model images)

Reviewer 2: @israel-hdez israel-hdez hey is it correct ?

Reviewer: The preview is rendering good for me.
Can you elaborate on your question?

Author: Per @mwaykole, we're all set here and I can mark this as resolved.

:_module-type: PROCEDURE

[id="using-oci-containers-for-model-storage_{context}"]
= Using OCI containers for model storage

[role='_abstract']

As an alternative to storing a model in an S3 bucket or URI, you can upload models to OCI containers. Using OCI containers for model storage can help you:

> * Reduce start up times by avoiding downloading the same model multiple times.

Author: ```suggestion
* Reduce startup times by avoiding downloading the same model multiple times.
```

* Reduce disk space usage by reducing the number of models downloaded locally.

> * Improve model performance by allowing pre-fetching images.

Reviewer 3: Should this be pre-fetched?

> This guide shows how to manually deploy a MobileNet v2-7 model in an ONNX format, stored in an OCI image on an OpenVINO model server.

Reviewer 3: [Optional] This guide shows **you** how to manually..

ifndef::upstream[]
[IMPORTANT]
====
ifdef::self-managed[]
Using OCI containers for model storage is currently available in {productname-long} {vernum} as a Technology Preview feature.
endif::[]
ifdef::cloud-service[]
Using OCI containers for model storage is currently available in {productname-long} as a Technology Preview feature.
endif::[]

> Technology Preview features are not supported with Red{nbsp}Hat production service level agreements (SLAs) and might not be functionally complete.

Reviewer 3: Suggestion: Replace Red {nbsp} Hat with the {org-name} attribute.

Author: ```suggestion
Technology Preview features are not supported with {org-name} production service level agreements (SLAs) and might not be functionally complete.
```

> Red{nbsp}Hat does not recommend using them in production.

Author: ```suggestion
{org-name} does not recommend using them in production.
```

These features provide early access to upcoming product features, enabling customers to test functionality and provide feedback during the development process.

For more information about the support scope of Red{nbsp}Hat Technology Preview features, see link:https://access.redhat.com/support/offerings/techpreview/[Technology Preview Features Support Scope].
====
endif::[]

.Prerequisites
* You have a model in the ONNX format.

== Creating an OCI image and storing a model in the container image

.Procedure
. From your local machine, create a temporary directory to store both the downloaded model and support files to create the OCI image:
+
[source]
----
cd $(mktemp -d)
----
+

> . Create a `models/` folder inside the temporary directory and download your model:

Author: ```suggestion
. Create a `models` folder inside the temporary directory and download your model:
```

====

== Deploying a model stored in an OCI image from a public repository

[NOTE]
====
By default in KServe, models are exposed outside the cluster and not protected with authorization. 
====

. Create a namespace to deploy the model:
+
[source]
----
oc new-project oci-model-example
----
+

ifndef::upstream[]
. Use the {productname-short} Applications project `kserve-ovms` template to create a `ServingRuntime` resource and configure the OpenVINO model server in the new namespace:
+
[source]
----
oc process -n redhat-ods-applications -o yaml kserve-ovms | oc apply -f -
----
endif::[]
ifdef::upstream[]
. Use the {productname-short} project `kserve-ovms` template to create a `ServingRuntime` resource and configure the OpenVINO model server in the new namespace:
+
[source]
----
oc process -n opendatahub -o yaml kserve-ovms | oc apply -f -
----
endif::[]
+
--
.. Verify that the `ServingRuntime` has been created with the `kserve-ovms` name:
+
[source]
----
oc get servingruntimes

NAME          DISABLED   MODELTYPE     CONTAINERS         AGE
kserve-ovms              openvino_ir   kserve-container   1m
----
--
+

> . Create an `InferenceService` YAML resource with the following values:

Reviewer 3: Should there be a command here to create the InferenceService? How is it created?

Author: @israel-hdez
Perhaps we should change to "Edit the `InferenceService`" ? IIRC, this resource is created when you set up model serving. Please keep me honest :)

Reviewer: @syaseen-rh is right. The YAML below represents the `InferenceServince` that needs to be created.

The YAML needs to be saved to a file and the command `oc apply -f {file.yaml}` will create the `InferenceService` in the cluster.

Author: Would something like this work then @israel-hdez ?
```suggestion
. Create an `InferenceService` YAML resource:
+
[source]
----
oc apply -f {file.yaml}
----
+
. Modify the `InferenceService` with the following values:
```

====

.Verification
After you create the `InferenceService` resource, KServe deploys the model stored in the OCI image referred to by the `storageUri` field. Check the status of the deployment with the following command:
[source]
----
oc get inferenceservice

NAME                    URL                                                       READY   PREV   LATEST   PREVROLLEDOUTREVISION   LATESTREADYREVISION                     AGE
sample-isvc-using-oci   https://sample-isvc-using-oci-oci-model-example.example   True           100                              sample-isvc-using-oci-predictor-00001   1m
----

== Deploying a model stored in an OCI image from a private repository

To deploy a model stored from a private OCI repository, you must configure an image pull secret. For more information about creating an image pull secret, see link:https://docs.openshift.com/container-platform/latest/openshift_images/managing_images/using-image-pull-secrets.html[Using image pull secrets^].

. Follow the steps in the previous section for deploying a model. However, when creating the `InferenceService` in step 3, specify your pull secret in the

> `spec.predictor.imagePullSecrets` field:

Reviewer 3: Minor comment, but wondering if you added the newline here unintentionally? Assuming it would be _specify your pull secret in the`spec.predictor.imagePullSecrets` field:_ , i.e, all in the same line.

Author: Great catch, thank you!
